src/gate/parser/bench/parser.cpp

```diff
@@ -0,0 +1,290 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the Utopia EDA Project, under the Apache License v2.0
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 ISP RAS (http://www.ispras.ru)
+//
+//===----------------------------------------------------------------------===//
+
+#include <FlexLexer.h>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <stdlib.h>
+#include <vector>
+
+#include "gate/model/gate.h"
+#include "gate/model/gnet.h"
+#include "gate/model/gsymbol.h"
+#include "headerFile"
+#include "tokens.h"
+
+extern "C" int scan_token();
+
+///Defines the compare method for TokenMap.name.
+#define COMPARE(text, line) do {\
+  if (yytext == i.name) {\
+    std::cout << "ERROR IN " << static_cast<const char*>(text) << " WITH "\
+      << yytext << "\tline: " << line << std::endl;\
+    exit(EXIT_FAILURE);\
+  }\
+} while (false)
+
+///Defines the alert method for debugging information.
+#define ALERT(errorString, line) do {\
+  std::cout << " expected '" << errorString << "'\tline: "\
+    << line << "\tСaught " << yytext << std::endl;\
+  exit(EXIT_FAILURE);\
+} while (false) 
+
+///Defines the method for verifying the next token.
+#define ASSERT_NEXT_TOKEN(expectedToken, errorString, place, line) do {\
+  Tokens token = getNextToken(place, line);\
+  if ( token != expectedToken ) {\
+    ALERT(errorString, line);\
+  }\
+} while (false)
+
+///Defines the adding in a token map vector.
+#define MAPS(definite, typeInit, maps) do {\
+  char* text = new char[strlen(yytext)];\
+  strncpy(text, yytext, strlen(yytext) + 1);\
+  TokenMap map {text, definite, typeInit};\
+  maps.push_back(map);\
+} while (false)
+
+using GNet = eda::gate::model::GNet;
+using Signal = eda::gate::model::Gate::Signal;
+using Gate = eda::gate::model::Gate;
+using GateSymbol = eda::gate::model::GateSymbol;
+
+struct TokenMap {
+  std::string name;
+  int definite; // 0 = NOT init, 1 = init
+  Tokens typeInit; // 0 = input, 1 = output, 2 = function
+  int usedInGate = 0;
+  // Gate::Id gateId;
+};
+
+void unknown(std::vector<TokenMap>& maps) {
+  std::vector<TokenMap> mapsUpd;
+  for (auto i : maps) {
+    for (auto j : maps) {
+      if (i.definite == 0 && j.definite == 1 && i.name == j.name) {
+          i.definite = 1;
+          mapsUpd.push_back(i);
+      }
+    }
+  }
+  for (auto i : maps) {
+    if(i.definite == 1)
+    mapsUpd.push_back(i);
+  }
+  for (auto i : mapsUpd) {
+    if (size(mapsUpd) != size(maps)) {
+      std::cout << "ERROR IN UNKNOWN DEFINITION " << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
```

Comment:
So, why it is in the loop?
```suggestion
  if (size(mapsUpd) != size(maps)) {
    std::cout << "ERROR IN UNKNOWN DEFINITION " << std::endl;
    exit(EXIT_FAILURE);
  }
```